kernel/src/util/ring_buffer.rs

```diff
@@ -0,0 +1,569 @@
+    ///
+    /// Returns an error if the ring buffer is full or
+    /// the underlying memory write failed.
+    pub fn push(&mut self, item: T) -> Result<()> {
+        let rb = &self.rb;
+        if rb.is_full() {
+            return_errno_with_message!(Errno::ENOMEM, "ring buffer is full")
+        }
+
+        let write_index = rb.write_index();
+        debug_assert!(write_index < rb.cap);
+
+        let segment_index = write_index * Self::T_SIZE;
+        let mut writer = rb.segment.writer().skip(segment_index).limit(Self::T_SIZE);
+        writer.write_val(&item)?;
+
+        rb.advance_write_index(write_index, 1);
+        Ok(())
+    }
+
+    /// Pushes a slice of items to the [`RingBuffer`].
+    ///
+    /// On success, all items are pushed to the ring buffer.
+    pub fn push_slice(&mut self, items: &[T]) -> Result<()> {
+        let nitems = items.len();
+        let rb = &self.rb;
+        let free_len = rb.free_len();
+        if free_len < nitems {
+            return_errno_with_message!(Errno::ENOMEM, "ring buffer is full or cannot fit all items")
+        }
+
+        let write_index = rb.write_index();
+        debug_assert!(write_index < rb.cap);
+        let segment_index = write_index * Self::T_SIZE;
+
+        if write_index + nitems > rb.cap {
+            // Write into two separate parts
+            rb.segment
+                .write_slice(segment_index, &items[..rb.cap - write_index])?;
+            rb.segment.write_slice(0, &items[rb.cap - write_index..])?;
+        } else {
+            rb.segment.write_slice(segment_index, items)?;
+        }
+
+        rb.advance_write_index(write_index, nitems);
+        Ok(())
+    }
+}
+
+impl<R: Deref<Target = RingBuffer<u8>>> Producer<u8, R> {
+    /// Writes data from the [`VmReader`] to the [`RingBuffer`].
+    ///
+    /// Returns the number of bytes written.
+    pub fn write(&mut self, reader: &mut VmReader) -> usize {
+        let rb = &self.rb;
+
+        let free_len = rb.free_len();
+        if free_len == 0 {
+            return 0;
+        }
+        let write_len = reader.remain().min(free_len);
+
+        let write_index = rb.write_index();
+        let write_len = if write_index + write_len > rb.cap {
+            // Write into two separate parts
+            let mut writer = rb
+                .segment
+                .writer()
+                .skip(write_index)
+                .limit(rb.cap - write_index);
+            let mut len = match writer.write_fallible(reader) {
+                Ok(len) => len,
+                Err((_, len)) => {
+                    rb.advance_write_index(write_index, len);
+                    return len;
+                }
+            };
+
+            let mut writer = rb
+                .segment
+                .writer()
+                .limit(write_len - (rb.cap - write_index));
+            len += match writer.write_fallible(reader) {
+                Ok(len) => len,
+                Err((_, len)) => len,
```

Comment:
You know there is a Rust feature called [try_blocks](https://doc.rust-lang.org/stable/unstable-book/language-features/try-blocks.html). Maybe you can use it to simply the error handling in this function.